### ISSUE TITLE
URL Path Traversal Vulnerability

### ISSUE THREAD

**Issue by AmiArt** [treefrogframework/treefrog-framework#142]
Found serious bug in TreeFrog. By adding ../ to URL path the attacker can view any file on server.

Example (reveals TreeFrog database config):
`curl -v -i --path-as-is http://localhost:8800/../config/database.ini`

![pic1](https://cloud.githubusercontent.com/assets/9396016/25938486/a521dbfa-362f-11e7-9ce7-fd4172ed8f78.png)

This happens because in tactioncontext.cpp /public path is concatenated with URL path without any validation:
![pic2](https://cloud.githubusercontent.com/assets/9396016/25938699/735b8b10-3630-11e7-9268-8c8d1bd3f144.png)

See also:
https://www.owasp.org/index.php/Path_Traversal

**Comment by treefrogframework** [comment]
Thank you for reporting.
Fix it. 29b6e1b9dcd7572f7e7400e5344f867eaafb90b2

**Comment by AmiArt** [comment]
On Windows QFileInfo::canonicalPath() replaces \ with / , so contains() always return false:

```
QFile reqPath(Tf::app()->publicPath() + path);
QFileInfo fi(reqPath);
 
if (fi.isFile() && fi.isReadable() && (fi.canonicalPath() + '/').contains(Tf::app()->publicPath())) {
    bool sendfile = true;
    // ...
}

// Tf::app()->publicPath()  is "C:/Users/Administrator/Documents/QtProjects\public\"
// fi.canonicalPath()       is "C:/Users/Administrator/Documents/QtProjects/public"
```

Why not validate path with QUrl class ?
And maybe a better place for URL validation is THttpRequestHeader ?

**Comment by treefrogframework** [comment]
> Why not validate path with QUrl class ?

Which method?
Could you send a pull-request?


Fix it.
1d8ae5b10cf1d7884eefc75629d05964597c292d

**Comment by AmiArt** [comment]
Still not fixed, how about (can't verify this, don't have compiled at hand):

QString canonicalPublicPath = QFileInfo(Tf::app()->publicPath()).canonicalPath();
if (canonicalPublicPath.endsWith('/') == false)
    canonicalPublicPath += '/';

if (fi.isFile() && fi.isReadable() && fi.canonicalFilePath().startsWith(canonicalPublicPath)) {
  // ...
}

PS. I was thinking about QUrl::resolved().

**Comment by AmiArt** [comment]
Fixed in b4a6f6b, thanks.
